docs/content/docs/concepts/Breaking Changes and Supported Patterns.md

```diff
@@ -0,0 +1,293 @@
+The below can be considered an initial draft for what some of the rules might look like.
+
+## Version Type Variance
+
+### An Example: Modifying interfaces
+
+```typescript
+interface Name {
+  first: string;
+  last?: string;
+}
+```
+
+Here TypeScript is doing a lot for us.
+It defines what values are valid as a `Named`, and what types can be used as a `Named`.
+This isn't enough in practice to define a useful Supported Pattern for users of this interface.
+Consider the two following examples of User Code for it:
+
+```typescript
+// User Code 1
+const myName1: Name = { first: "My" };
+// User Code 2
+const myName2: Name = { first: "My", last: "Name" };
+// User Code 3
+interface FullName extends Name {
+  // Some people have multiple middle names, or none.
+  middle: string[];
+}
+```
+
+What ways can the package change `Named` without breaking any of these users?
+
+```typescript
+interface Name {
+  first: string;
+  last: string; // Make required. Causes User Code 1 to not build.
+}
+```
+
+```typescript
+interface Name {
+  first: string;
+  middle?: string; // Added optional field. Causes User Code 3 to not build.
+  last: string;
+}
+```
+
+```typescript
+interface Name {
+  first: string;
+  // Removed `last`. Causes User Code 2 to not build.
+}
+```
+
+### The Rule
+
+As shown in the example above, even when only consider if the code builds, its impossible to add or remove fields, optional or otherwise, or adjust the optionality of any field on an interface as a backwards compatible change if all three of these patterns are supported.
+In general, unless the supported patterns for using an interface is restricted to not permit these kinds of usage, modifying it in a compatible way is impossible.
+
+There are a few approaches that can be taken to resolve this:
+
+1. Restrict the Supported Patterns for the interface. Depending on how its used there are two main choices:
+   1. "out": Disallow implementing and extending the interface and modifying instances. This means the package can add members, remove optionality and make members have more specific types. This means that the old interface must be assignable to the new interface. Uses also should not assume that enumerating members of objects that meet the interface won't return unexpected entries (this is true in general, but extra important for interfaces using these rules). This restricts the instances of the interface to being produced by the package making it logically an "output" of the package.
+   2. "in": Disallow reading members of the interface (except via `{added: value, ...existingInstance}` pattern). This means the package can remove members, add optionality and make members have more general types. This means that the new interface must be assignable to the old interface. This restricts the instances of the interface to being build but not consumed by user code making it logically an "input" of the package.
```

Comment:
Am I reading "interface X must be assignable to interface Y" backwards, or are they reversed for the "in" and "out" cases? For "out", adding a member would mean that the old interface cannot be used where the new one is required, but "the old interface must be assignable to the new interface" makes me think this of this, which wouldn't work:
```typescript
interface OldInterface {
  name: string;
}
interface NewInterface {
  name: string,
  lastName: string;
}
const old: OldInterface = { name: "John" };
const new: NewInterface = old;
```

Reply:
Yes, I got that backwards.

Reply:
```suggestion
   1. "out": Disallow implementing and extending the interface and modifying instances. This means the package can add members, remove optionality and make members have more specific types. This means that the new interface must be assignable to the old interface. Uses also should not assume that enumerating members of objects that meet the interface won't return unexpected entries (this is true in general, but extra important for interfaces using these rules). This restricts the instances of the interface to being produced by the package making it logically an "output" of the package.
   2. "in": Disallow reading members of the interface (except via `{added: value, ...existingInstance}` pattern). This means the package can remove members, add optionality and make members have more general types. This means that the old interface must be assignable to the new interface. This restricts the instances of the interface to being build but not consumed by user code making it logically an "input" of the package.
```
That should fix it.

It's hard to get this right because TypeScripts rules are unsound makes this kind of reasoning hard (ex: I'd use readonly here, but typescript choose to ignore it for assignability of interfaces).